qiskit/ignis/experiments/calibration/instruction_data/compiler.py

```diff
@@ -0,0 +1,318 @@
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Compile calibration schedule DSL to pulse Schedule."""
+
+import dataclasses
+import re
+from enum import Enum
+from typing import List, Dict, NamedTuple, Iterator, Union, Optional
+
+from qiskit import pulse, circuit
+from qiskit.ignis.experiments.calibration.instruction_data import utils
+
+
+class TokenSpec(Enum):
+    """Map the AST node name to the DSL syntax.
+
+    The enum name is the AST node type, the value is its mapping to the DSL syntax.
+    """
+    CONTEXT_ENTER = r'\[(?P<pos>right|left|seq)\]{'
+    CONTEXT_EXIT = r'}'
+    REFERENCE = r'&(?P<name>[\w]*)'
+    PULSE = r'%(?P<name>[\w]*)\((?P<channel>[a-z][0-9]+),(?P<shape>[\w]*)\)'
+    FRAME = r'$(?P<name>[\w]*)\((?P<channel>[a-z][0-9]+),(?P<operand>[0-9.]+)\)'
+
+
+#
+# Create token from input string
+#
+
+Token = NamedTuple('Token', [('type', str), ('data', Dict[str, str])])
+
+
+def _tokenizer(source: str) -> Iterator[Token]:
+    """A helper function to generate token from code string.
+
+    Args:
+        source: Input calibration schedule DSL.
+
+    Returns:
+        Iterator of token object.
+    """
+    source = source.replace(' ', '')
+    while source:
+        for spec in TokenSpec:
+            matched_obj = re.match(spec.value, source)
+            if matched_obj:
+                source = source[matched_obj.end():]
+                token = Token(spec.name, matched_obj.groupdict())
+                break
+        else:
+            raise Exception('Invalid syntax {source}'.format(source=source))
+        yield token
+
+
+#
+# Create AST from token list
+#
+
+@dataclasses.dataclass(frozen=True)
+class Inst:
+    """A base class of instruction. This class is meant to be subclassed."""
+    name: str
+
+
+@dataclasses.dataclass(frozen=True)
+class PulseInst(Inst):
+    """A leaf that represents a play schedule component."""
+    channel: str
+    shape: str
+
+    def __repr__(self):
+        return 'Pulse({}.{}.{})'.format(self.name, self.channel, self.shape)
+
+
+@dataclasses.dataclass(frozen=True)
+class FrameInst(Inst):
+    """A leaf that represents non-pulse schedule component such as frame change."""
+    channel: str
+    operand: float
+
+    def __repr__(self):
+        return 'Frame({},{})'.format(self.name, self.operand)
+
+
+@dataclasses.dataclass(frozen=True)
+class Reference(Inst):
+    """A leaf that represents a reference to another schedule definition."""
+
+    def __repr__(self):
+        return 'Ref({})'.format(self.name)
+
+
+@dataclasses.dataclass
+class ScheduleBlock:
+    """A node that represents a block of schedule components."""
```

Comment:
Should we somehow specify that this is a left-aligned block?

Reply:
By default. This depends on the schedule code.

Reply:
```suggestion
    """A node that represents a block of schedule components, left-aligned by default."""
```